src/realm/sort_descriptor.cpp

```diff
@@ -23,9 +23,52 @@
 #include <realm/util/assert.hpp>
 #include <realm/list.hpp>
 #include <realm/dictionary.hpp>
+#include <realm/string_interner.hpp>
 
 using namespace realm;
 
+namespace {
+
+template <typename T, typename Col>
+int compare(const T& i, const T& j, const Col& col)
+{
+    Mixed m_i = i.get_value();
+    Mixed m_j = j.get_value();
+
+    // 1. not compressed
+    if (!i.compressed && !j.compressed)
+        return m_i.compare(m_j);
+
+    ColKey ck{col.col_key};
+    StringInterner* interner = col.table->get_string_interner(ck);
+
+    // 2. two compressed strings
+    if (i.compressed && j.compressed) {
+        return interner->compare((StringID)m_i.get_int(), (StringID)m_j.get_int());
+    }
+
+    // 3. one index is a compressed string, and the other one is mixed.
+    if (i.compressed || j.compressed) {
+        if (m_i.is_type(type_String))
+            return interner->compare(m_i.get_string(), (StringID)m_j.get_int());
+
+        if (m_j.is_type(type_String))
+            return -interner->compare(m_j.get_string(), (StringID)m_i.get_int());
+    }
+
+    // 4. compare string vs any other non-string (this is going to be slow)
```

Comment:
if one side is a string and the other isn't then we don't actually need to fetch the string value because the Mixed comparison is only about the type ordering, not the values. So you could speed this up by just using a constant empty string.